server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -710,9 +714,27 @@ static boolean shouldPreFilterSearchShards(ClusterState clusterState,
         } else if (preFilterShardSize == null) {
             preFilterShardSize = SearchRequest.DEFAULT_PRE_FILTER_SHARD_SIZE;
         }
+        if (RollupV2.isEnabled() && hasRollupDatastream(indices, searchRequest.indices(), clusterState)) {
+            return true;
+        }
         return searchRequest.searchType() == QUERY_THEN_FETCH // we can't do this for DFS it needs to fan out to all shards all the time
-                    && (SearchService.canRewriteToMatchNone(source) || hasPrimaryFieldSort(source))
-                    && preFilterShardSize < numShards;
+            && (SearchService.canRewriteToMatchNone(source) || hasPrimaryFieldSort(source))
+            && preFilterShardSize < numShards;
+    }
+
+    private static boolean hasRollupDatastream(String[] indices, String[] requestIndices, ClusterState clusterState) {
```

Comment:
That's not the logic that we need imo.
We want to know if a datastream is explicitly requested so it should use`SearchRequest#indices` to match datastreams ?